### PR TITLE
add framework for estimator server

### LINKED ISSUE #277: Implement Built-in Default Replica Predictor
- [x] #336 
- [ ] #346

### PR DESCRIPTION
**Pull request by qianjun1993**
#### What type of PR is this?
kind/feature

#### What this PR does / why we need it:

estimator has four steps , and we want to make each of them easy to extend

#### Which issue(s) this PR fixes:

Fixes #277 

#### Special notes for your reviewer:
only the framework logic is in this pr. 
Other logic of default estimator will be in other PR

### CODE REVIEW

File: pkg/estimator/apis/types.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/*
```
Comment (dixudx) [inline pkg/estimator/apis/types.go:1]: Please move to `pkg/apis/scheduler`.

Comment (qianjun1993) [reply pkg/estimator/apis/types.go:1]: framework of scheduler is also defined in pkg/scheduler/apis/types.

**Review by dixudx — APPROVED** [review]

File: pkg/estimator/apis/types.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	Filter PluginSet
+
+	// PostFilter is a list of plugins that are invoked after filtering phase, no matter whether filtering succeeds or not.
+	PostFilter PluginSet
+
+	// PreCalculate is a list of plugins that are invoked before Calculating.
+	PreCalculate PluginSet
+
+	// Calculate is a list of plugins that should be invoked when calculate max available replicas for nodes that have passed the filtering phase.
+	Calculate PluginSet
```
Comment (Garrybest) [inline pkg/estimator/apis/types.go:44]: What about `Compute PluginSet`?

File: pkg/estimator/framework/interfaces/interface.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2021 The Clusternet Authors.
```
Comment (Garrybest) [inline pkg/estimator/framework/interfaces/interface.go:2]: 2022

**Comment by silenceper** [comment]
Each estimator is actually equivalent to the filter stage in kube-scheduler?

**Comment by dixudx** [comment]
> Each estimator is actually equivalent to the filter stage in kube-scheduler?
> 
> I see from here (#322 )that the estimate phase only cares about the number of replicas

This framework is for server side implementation of estimator, which runs with `clusternet-agent`.
`kube-scheduler` will send HTTP requests to estimator server.

File: pkg/estimator/apis/types.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	// PreScore is a list of plugins that are invoked before scoring.
+	PreScore PluginSet
+
+	// Score is a list of plugins that should be invoked when ranking clusters that have passed the filtering phase.
+	Score PluginSet
```
Comment (silenceper) [inline pkg/estimator/apis/types.go:50]: why we need score plugin, when would this stage be required?

File: pkg/estimator/apis/types.go (hunk 2 above)
Comment (qianjun1993) [reply pkg/estimator/apis/types.go]: update it to Compute

File: pkg/estimator/framework/interfaces/interface.go (hunk 3 above)
Comment (qianjun1993) [reply pkg/estimator/framework/interfaces/interface.go]: update it to 2022

**Review by dixudx — APPROVED** [review]

**Pull request merged**